tests/ignite/engine/test_engine.py

```diff
@@ -124,32 +124,147 @@ def start_of_epoch_handler(engine):
+
+            expected_iter += 1
+            assert engine.state.iteration == expected_iter
+
+            try:
+                assert engine.state.batch == next(expected_data_iter)
+            except StopIteration:
+                expected_data_iter = iter(data)
+                assert engine.state.batch == next(expected_data_iter)
+
+    first_epoch_iter[0], first_epoch_iter[1] = state.epoch, state.iteration
+    state = engine.run(data, max_epochs=max_epochs, epoch_length=epoch_length)
+
+    assert state.epoch == max_epochs
+    assert not engine.should_terminate
+    # As terminated epoch is skipped -> iterations are not incremented
+    assert state.iteration == real_epoch_length * (max_epochs - 1)
+
+
+@pytest.mark.parametrize("data, epoch_length", [(None, 10), (range(10), None)])
+def test_terminate_stops_run_mid_epoch(data, epoch_length):
+    max_epochs = 5
+    iteration_to_stop = 13
+    real_epoch_length = epoch_length if data is None else len(data)
 
     engine = Engine(MagicMock(return_value=1))
 
     def start_of_iteration_handler(engine):
         if engine.state.iteration == iteration_to_stop:
             engine.terminate()
 
+    @engine.on(Events.EXCEPTION_RAISED)
+    def assert_no_exceptions(ee):
+        assert False, f"Engine should terminate without raising an exception, got '{type(ee)}'"
+
     engine.add_event_handler(Events.ITERATION_STARTED, start_of_iteration_handler)
-    state = engine.run(data, max_epochs=3, epoch_length=num_iterations_per_epoch)
+    state = engine.run(data, max_epochs=max_epochs, epoch_length=epoch_length)
     # completes the iteration but doesn't increment counter (this happens just before a new iteration starts)
     assert state.iteration == iteration_to_stop
-    assert state.epoch == np.ceil(iteration_to_stop / num_iterations_per_epoch)  # it starts from 0
+    assert state.epoch == np.ceil(iteration_to_stop / real_epoch_length)  # it starts from 0
+
+    # Engine continue from epoch_to_terminate_on until max_epochs
+    first_epoch_iter = [None, None]
+
+    @engine.on(Events.STARTED, first_epoch_iter)
+    def check_iter_epoch(first_epoch_iter):
+        assert engine.state.epoch == first_epoch_iter[0]
+        assert engine.state.iteration == first_epoch_iter[1]
+
+    if data is not None:
+        expected_iter = state.iteration
+
+        @engine.on(Events.ITERATION_STARTED)
+        def check_iter_and_data():
+            nonlocal expected_iter
+
+            expected_iter += 1
+            assert engine.state.iteration == expected_iter
+            assert engine.state.batch == data[(expected_iter - first_epoch_iter[1] - 1) % len(data)]
+
+    first_epoch_iter[0], first_epoch_iter[1] = state.epoch, state.iteration
+    state = engine.run(data, max_epochs=max_epochs, epoch_length=epoch_length)
+
+    assert state.epoch == max_epochs
+    assert not engine.should_terminate
+    assert state.iteration == real_epoch_length * (max_epochs - 1)
```

Comment:
This shouldn't be `state.iteration == real_epoch_length * (max_epochs - 1) + (iteration_to_stop % real_epoch_length)` ? Since engine has run a few iterations from the epoch in which it stopped.

Reply:
As this test is passing then no, it is `state.iteration == real_epoch_length * (max_epochs - 1)`. 
However, I agree that we may expect iteration to be `state.iteration == real_epoch_length * (max_epochs - 1) + (iteration_to_stop % real_epoch_length)`...

Right now if we inspect epoch/iteration and data attributes we have the following:
```
- (epoch, iter, data): 1 1  |  0
- (epoch, iter, data): 1 2  |  1
- (epoch, iter, data): 1 3  |  2
- (epoch, iter, data): 1 4  |  3
- (epoch, iter, data): 1 5  |  4
- (epoch, iter, data): 1 6  |  5
- (epoch, iter, data): 1 7  |  6
- (epoch, iter, data): 1 8  |  7
- (epoch, iter, data): 1 9  |  8
- (epoch, iter, data): 1 10  |  9
- (epoch, iter, data): 2 11  |  0
- (epoch, iter, data): 2 12  |  1
- (epoch, iter, data): 2 13  |  2
-> terminate
Continue the run: 2 13
- (epoch, iter, data): 3 14  |  0
- (epoch, iter, data): 3 15  |  1
- (epoch, iter, data): 3 16  |  2
- (epoch, iter, data): 3 17  |  3
- (epoch, iter, data): 3 18  |  4
- (epoch, iter, data): 3 19  |  5
- (epoch, iter, data): 3 20  |  6
- (epoch, iter, data): 4 21  |  7
- (epoch, iter, data): 4 22  |  8
- (epoch, iter, data): 4 23  |  9
- (epoch, iter, data): 4 24  |  0
- (epoch, iter, data): 4 25  |  1
- (epoch, iter, data): 4 26  |  2
- (epoch, iter, data): 4 27  |  3
- (epoch, iter, data): 4 28  |  4
- (epoch, iter, data): 4 29  |  5
- (epoch, iter, data): 4 30  |  6
- (epoch, iter, data): 5 31  |  7
- (epoch, iter, data): 5 32  |  8
- (epoch, iter, data): 5 33  |  9
- (epoch, iter, data): 5 34  |  0
- (epoch, iter, data): 5 35  |  1
- (epoch, iter, data): 5 36  |  2
- (epoch, iter, data): 5 37  |  3
- (epoch, iter, data): 5 38  |  4
- (epoch, iter, data): 5 39  |  5
- (epoch, iter, data): 5 40  |  6
```

Does it look strange or weird to you ?

Reply:
Epoch 3 is supposed to be run over all data not the first 7 batches:
![image](https://user-images.githubusercontent.com/22097587/186178466-0af482a3-e3fe-4859-8972-f2c0a07d76ba.png)

Reply:
It's due to `iter_counter` local variable (in `run_once_on_dataset`) which has been initialized as `state.iteration % epoch_length` (3)  and causes engine to stop 7 batches later when `iter_counter == epoch_length`.

Hence, `iter_counter` should start with zero.

Reply:
Let's move the discussions in this issue: https://github.com/pytorch/ignite/issues/2669